### PR TITLE
fix(`launch`): add check for MsgSettleRequest simulation

### LINKED ISSUE #405: Simulation test: `MsgSettleRequest` verify that the request can be apply before approving it
https://github.com/tendermint/spn/blob/3d0d523cd21296202d9dcea66e43f623cc7f40f9/x/launch/simulation/simulation.go#L362

There is no verification if the request can be applied before sending the message (eg: the genesis account already exists)
If the request can't be applied, the message fails and the simulation stops
We should check if the request can be applied and send a `NoOpMsg ` if not (or reject the request)

### PR DESCRIPTION
**Pull request by aljo242**
Closes #405

Perform a check using the logic of `ApplyRequest()` without modifying state.  Implemented in a new function in the simulation package called `checkRequest()`.

### CODE REVIEW

**Comment by giunatale** [comment]
Is this fixing the current simulation error on `develop`?

**Comment by aljo242** [comment]
> Is this fixing the current simulation error on `develop`?

Yessir :)

**Review by giunatale — CHANGES_REQUESTED** [review]
small lint fix, otherwise LGTM

File: x/launch/simulation/store.go [hunk 1]
```diff
@@ -2,6 +2,8 @@ package simulation
 
 import (
 	"fmt"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
```
[the remark this answers is not in the document]
Comment (giunatale) [reply x/launch/simulation/store.go:7]: we should keep this format here, where imports are grouped with spn at the bottom. That's what I've seen done everywhere.

**Comment by lubtd** [comment]
Can you try the simulation with some different seed to ensure if we no longer get this error?

```
starport chain simulate --seed [nb]
```

**Comment by aljo242** [comment]
> 

Ran with seeds [1, 15] and everything is ok.

File: x/launch/simulation/store.go [hunk 2]
```diff
@@ -163,3 +165,76 @@ func FindRandomValidator(
 	}
 	return simAccount, valAcc, found
 }
+
+func checkRequest(
```
[the remark this answers is not in the document]
Comment (aljo242) [reply x/launch/simulation/store.go]: Ok, I removed the check part and created `CheckRequest` in the keeper package. This means we perform the switch-case on the request contents twice (once to check, once to apply), but the code is a bit cleaner now i think.

**Comment by aljo242** [comment]
Ok, I removed the check part and created CheckRequest in the keeper package. This means we perform the switch-case on the request contents twice (once to check, once to apply), but the code is a bit cleaner now i think.

**Comment by lubtd** [comment]
> Ok, I removed the check part and created CheckRequest() in the keeper package. This means we perform the switch-case on the request contents twice (once to check, once to apply), but the code is a bit cleaner now i think.

This is fine

File: x/launch/keeper/request.go [hunk 3]
```diff
@@ -133,6 +133,40 @@ func ApplyRequest(
...
+		vr := requestContent.ValidatorRemoval
+		k.RemoveGenesisValidator(ctx, launchID, vr.ValAddress)
+	default:
+		return spnerrors.Critical("unknown request content type")
+	}
+	return nil
+}
+
+// CheckRequest verifies that a request can be applied
+func CheckRequest(
```
Comment (lubtd) [inline x/launch/keeper/request.go:165]: We should create the `TestCheckRequest` test to test request checking

`TestApplyRequest` would keep the state transition assertion and should verify that it fails if a request is invalid (without testing all invalid requests)

**Review by giunatale — DISMISSED** [review]

File: x/launch/keeper/request_test.go [hunk 4]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+	)
+	tests := []struct {
+		name    string
+		request types.Request
+		err     error
+	}{
+		{
+			name:    "test GenesisAccount content",
+			request: sample.RequestWithContent(launchID, contents[0]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:206]:
```suggestion
```

File: x/launch/keeper/request_test.go [hunk 5]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+		}, {
+			name:    "test duplicated GenesisAccount content",
+			request: sample.RequestWithContent(launchID, contents[0]),
+			err:     types.ErrAccountAlreadyExist,
+		}, {
+			name:    "test genesis AccountRemoval content",
+			request: sample.RequestWithContent(launchID, contents[1]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:214]:
```suggestion
```

File: x/launch/keeper/request_test.go [hunk 6]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+		}, {
+			name:    "test not found genesis AccountRemoval content",
+			request: sample.RequestWithContent(launchID, contents[1]),
+			err:     types.ErrAccountNotFound,
+		}, {
+			name:    "test VestingAccount content",
+			request: sample.RequestWithContent(launchID, contents[2]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:222]:
```suggestion
```

File: x/launch/keeper/request_test.go [hunk 7]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+		}, {
+			name:    "test duplicated VestingAccount content",
+			request: sample.RequestWithContent(launchID, contents[2]),
+			err:     types.ErrAccountAlreadyExist,
+		}, {
+			name:    "test vesting AccountRemoval content",
+			request: sample.RequestWithContent(launchID, contents[3]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:230]:
```suggestion
```

File: x/launch/keeper/request_test.go [hunk 8]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+		}, {
+			name:    "test not found vesting AccountRemoval content",
+			request: sample.RequestWithContent(launchID, contents[3]),
+			err:     types.ErrAccountNotFound,
+		}, {
+			name:    "test GenesisValidator content",
+			request: sample.RequestWithContent(launchID, contents[4]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:238]:
```suggestion
```

File: x/launch/keeper/request_test.go [hunk 9]
```diff
@@ -183,3 +184,85 @@ func TestApplyRequest(t *testing.T) {
...
+		}, {
+			name:    "test duplicated GenesisValidator content",
+			request: sample.RequestWithContent(launchID, contents[4]),
+			err:     types.ErrValidatorAlreadyExist,
+		}, {
+			name:    "test ValidatorRemoval content",
+			request: sample.RequestWithContent(launchID, contents[5]),
+			err:     nil,
```
Comment (lubtd) [inline x/launch/keeper/request_test.go:246]:
```suggestion
```

**Review by lubtd — APPROVED** [review]

**Pull request merged**